srml/contracts/src/account_db.rs

```diff
@@ -225,7 +218,7 @@ impl<'a, T: Trait> OverlayAccountDb<'a, T> {
 		contract.code_hash = Some(code_hash);
 		contract.rent_allowance = Some(<BalanceOf<T>>::max_value());
 
-		Ok(())
+		Ok(new_contract_info::<T>(account, code_hash))
```

Comment:
This generate a new trie_id which is temporary, it will only be used for creating the nested execution environment so it will only be used for get values and as it is empty this is fine. When committing to DirectAccountDB it will create another TrieId.

@pepyakin is there a reason not to generate a temporary unused trie_id ?

also with the upcoming https://github.com/paritytech/substrate/pull/2209 creating a trie_id will actually create an empty Trie I think, which allocate some ressource (trie extension and maybe some other stuff).
also it needs to be written in `COMPLEXITY.md` that that create_contract will do a increase AccountCounter. (also maybe it should be written for commit as well)

Reply:
yeah, this catched my eye as well. The problem is that `trie_id` call is not cost-free, since it does mutation of a `AccountCounter` and performs some hashing.

Good point about #2209

Reply:
Ah, I forgot `TrieIdGenerator` impl is stateful.